openedx/core/djangoapps/course_groups/cohorts.py

```diff
@@ -244,24 +244,19 @@ def get_cohort(user, course_key, assign=True, use_cached=False):
 
     # Otherwise assign the user a cohort.
     try:
-        with transaction.atomic():
-            # If learner has been pre-registered in a cohort, get that cohort. Otherwise assign to a random cohort.
-            course_user_group = None
-            for assignment in UnregisteredLearnerCohortAssignments.objects.filter(email=user.email, course_id=course_key):
-                course_user_group = assignment.course_user_group
-                unregistered_learner = assignment
-
-            if course_user_group:
-                unregistered_learner.delete()
-            else:
-                course_user_group = get_random_cohort(course_key)
-
-            membership = CohortMembership.objects.create(
-                user=user,
-                course_user_group=course_user_group,
-            )
-
-            return cache.setdefault(cache_key, membership.course_user_group)
+        # If learner has been pre-registered in a cohort, get that cohort. Otherwise assign to a random cohort.
+        course_user_group = None
+        for assignment in UnregisteredLearnerCohortAssignments.objects.filter(email=user.email, course_id=course_key):
+            course_user_group = assignment.course_user_group
+            assignment.delete()
+            break
+        else:
+            course_user_group = get_random_cohort(course_key)
+        add_user_to_cohort(course_user_group, user)
+        return course_user_group
+    except ValueError:
```

Comment:
What raises the `ValueError`?

Reply:
The doc string to add_user_to_cohort says it'll raise a ``ValueError`` if the user is already in the cohort.